<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%205/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Content Based Filtering

Estimated time needed: **25** minutes

## Objectives

After completing this lab you will be able to:

*   Create a recommendation system using Content Based filtering


Recommendation systems are a collection of algorithms used to recommend items to users based on information taken from the user. These systems have become ubiquitous, and can be commonly seen in online stores, movies databases and job finders. In this notebook, we will explore Content-based recommendation systems and implement a simple version of one using Python and the Pandas library.


### Table of contents

<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ol>
        <li><a href="https://#ref1">Acquiring the Data</a></li>
        <li><a href="https://#ref2">Preprocessing</a></li>
        <li><a href="https://#ref3">Content-Based Filtering</a></li>
    </ol>
</div>
<br>


<a id="ref1"></a>

# Acquiring the Data


To acquire and extract the data, simply run the following Bash scripts:\
Dataset acquired from [GroupLens](http://grouplens.org/datasets/movielens/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01). Let's download the dataset. To download the data, we will use **`!wget`** to download it from IBM Object Storage.\
**Did you know?** When it comes to Machine Learning, you will likely be working with large datasets. As a business, where can you host your data? IBM is offering a unique opportunity for businesses, with 10 Tb of IBM Cloud Object Storage: [Sign up now for free](http://cocl.us/ML0101EN-IBM-Offer-CC)


In [4]:
!wget -O moviedataset.zip https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%205/data/moviedataset.zip
print('unzipping ...')
!unzip -o -j moviedataset.zip 

unzipping ...


--2022-05-04 15:35:26--  https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%205/data/moviedataset.zip
Resolving cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)... 169.45.118.108
Connecting to cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)|169.45.118.108|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 160301210 (153M) [application/zip]
Saving to: 'moviedataset.zip'

     0K .......... .......... .......... .......... ..........  0% 96.1K 27m8s
    50K .......... .......... .......... .......... ..........  0%  195K 20m15s
   100K .......... .......... .......... .......... ..........  0% 6.31M 13m38s
   150K .......... .......... .......... .......... ..........  0%  209K 13m20s
   200K .......... .......... .......... .......... ..........  0

  6600K .......... .......... .......... .......... ..........  4% 11.5M 5m26s
  6650K .......... .......... .......... .......... ..........  4%  203M 5m23s
  6700K .......... .......... .......... .......... ..........  4%  300M 5m21s
  6750K .......... .......... .......... .......... ..........  4% 63.4K 5m36s
  6800K .......... .......... .......... .......... ..........  4%  265M 5m33s
  6850K .......... .......... .......... .......... ..........  4%  346M 5m31s
  6900K .......... .......... .......... .......... ..........  4%  268M 5m28s
  6950K .......... .......... .......... .......... ..........  4%  201K 5m31s
  7000K .......... .......... .......... .......... ..........  4%  235M 5m29s
  7050K .......... .......... .......... .......... ..........  4%  182M 5m26s
  7100K .......... .......... .......... .......... ..........  4% 4.41M 5m24s
  7150K .......... .......... .......... .......... ..........  4%  130M 5m22s
  7200K .......... .......... .......... .......... 

 27100K .......... .......... .......... .......... .......... 17%  170K 8m19s
 27150K .......... .......... .......... .......... .......... 17%  187K 8m19s
 27200K .......... .......... .......... .......... .......... 17%  118K 8m20s
 27250K .......... .......... .......... .......... .......... 17% 12.6M 8m19s
 27300K .......... .......... .......... .......... .......... 17% 2.12M 8m18s
 27350K .......... .......... .......... .......... .......... 17%  153K 8m18s
 27400K .......... .......... .......... .......... .......... 17%  182K 8m18s
 27450K .......... .......... .......... .......... .......... 17%  181K 8m18s
 27500K .......... .......... .......... .......... .......... 17% 8.61M 8m17s
 27550K .......... .......... .......... .......... .......... 17%  359M 8m16s
 27600K .......... .......... .......... .......... .......... 17%  202K 8m16s
 27650K .......... .......... .......... .......... .......... 17% 6.09M 8m15s
 27700K .......... .......... .......... .......... 

 41100K .......... .......... .......... .......... .......... 26%  542K 6m55s
 41150K .......... .......... .......... .......... .......... 26%  288K 6m55s
 41200K .......... .......... .......... .......... .......... 26%  559K 6m55s
 41250K .......... .......... .......... .......... .......... 26%  211K 6m55s
 41300K .......... .......... .......... .......... .......... 26%  257K 6m54s
 41350K .......... .......... .......... .......... .......... 26%  781K 6m54s
 41400K .......... .......... .......... .......... .......... 26%  264K 6m54s
 41450K .......... .......... .......... .......... .......... 26%  682K 6m53s
 41500K .......... .......... .......... .......... .......... 26%  274K 6m53s
 41550K .......... .......... .......... .......... .......... 26%  570K 6m53s
 41600K .......... .......... .......... .......... .......... 26%  284K 6m52s
 41650K .......... .......... .......... .......... .......... 26%  601K 6m52s
 41700K .......... .......... .......... .......... 

 81950K .......... .......... .......... .......... .......... 52% 1.48M 3m37s
 82000K .......... .......... .......... .......... .......... 52% 2.92M 3m37s
 82050K .......... .......... .......... .......... .......... 52% 2.83M 3m36s
 82100K .......... .......... .......... .......... .......... 52% 77.7K 3m37s
 82150K .......... .......... .......... .......... .......... 52%  189K 3m37s
 82200K .......... .......... .......... .......... .......... 52% 11.2M 3m36s
 82250K .......... .......... .......... .......... .......... 52%  209M 3m36s
 82300K .......... .......... .......... .......... .......... 52% 2.35M 3m36s
 82350K .......... .......... .......... .......... .......... 52% 4.18M 3m36s
 82400K .......... .......... .......... .......... .......... 52% 56.8K 3m36s
 82450K .......... .......... .......... .......... .......... 52%  250M 3m36s
 82500K .......... .......... .......... .......... .......... 52%  269M 3m36s
 82550K .......... .......... .......... .......... 

 94350K .......... .......... .......... .......... .......... 60% 4.24M 2m52s
 94400K .......... .......... .......... .......... .......... 60%  138K 2m52s
 94450K .......... .......... .......... .......... .......... 60%  288K 2m52s
 94500K .......... .......... .......... .......... .......... 60%  241M 2m52s
 94550K .......... .......... .......... .......... .......... 60%  305M 2m52s
 94600K .......... .......... .......... .......... .......... 60%  386M 2m51s
 94650K .......... .......... .......... .......... .......... 60% 2.79M 2m51s
 94700K .......... .......... .......... .......... .......... 60%  197K 2m51s
 94750K .......... .......... .......... .......... .......... 60%  218M 2m51s
 94800K .......... .......... .......... .......... .......... 60%  221M 2m51s
 94850K .......... .......... .......... .......... .......... 60% 11.8M 2m50s
 94900K .......... .......... .......... .......... .......... 60% 81.6K 2m51s
 94950K .......... .......... .......... .......... 

121600K .......... .......... .......... .......... .......... 77% 7.21M 97s
121650K .......... .......... .......... .......... .......... 77%  217M 97s
121700K .......... .......... .......... .......... .......... 77% 11.0M 97s
121750K .......... .......... .......... .......... .......... 77%  175K 96s
121800K .......... .......... .......... .......... .......... 77%  393M 96s
121850K .......... .......... .......... .......... .......... 77%  177K 96s
121900K .......... .......... .......... .......... .......... 77%  164M 96s
121950K .......... .......... .......... .......... .......... 77%  251M 96s
122000K .......... .......... .......... .......... .......... 77% 50.5K 96s
122050K .......... .......... .......... .......... .......... 77%  183K 96s
122100K .......... .......... .......... .......... .......... 78% 2.18M 96s
122150K .......... .......... .......... .......... .......... 78%  255M 95s
122200K .......... .......... .......... .......... .......... 78%  111M 95s

Now you're ready to start working with the data!


<a id="ref2"></a>

# Preprocessing


First, let's get all of the imports out of the way:


In [3]:
#Dataframe manipulation library
import pandas as pd
#Math functions, we'll only need the sqrt function so let's import only that
from math import sqrt
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

Now let's read each file into their Dataframes:


In [5]:
#Storing the movie information into a pandas dataframe
movies_df = pd.read_csv('movies.csv')
#Storing the user information into a pandas dataframe
ratings_df = pd.read_csv('ratings.csv')
#Head is a function that gets the first N rows of a dataframe. N's default is 5.
movies_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

Let's also remove the year from the **title** column by using pandas' replace function and store in a new **year** column.


In [7]:
#Using regular expressions to find a year stored between parentheses
#We specify the parantheses so we don't conflict with movies that have years in their titles
movies_df['year'] = movies_df.title.str.extract('(\(\d\d\d\d\))',expand=False)
#Removing the parentheses
movies_df['year'] = movies_df.year.str.extract('(\d\d\d\d)',expand=False)
#Removing the years from the 'title' column
movies_df['title'] = movies_df.title.str.replace('(\(\d\d\d\d\))', '')
#Applying the strip function to get rid of any ending whitespace characters that may have appeared
movies_df['title'] = movies_df['title'].apply(lambda x: x.strip())
movies_df.head()

C:\Users\admin\AppData\Local\Temp\ipykernel_18044\1143695627.py:7: FutureWarning: The default value of regex will change from True to False in a future version.
  movies_df['title'] = movies_df.title.str.replace('(\(\d\d\d\d\))', '')


movieId                        title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   

                                        genres  year  
0  Adventure|Animation|Children|Comedy|Fantasy  1995  
1                   Adventure|Children|Fantasy  1995  
2                               Comedy|Romance  1995  
3                         Comedy|Drama|Romance  1995  
4                                       Comedy  1995

With that, let's also split the values in the **Genres** column into a **list of Genres** to simplify for future use. This can be achieved by applying Python's split string function on the correct column.


In [8]:
#Every genre is separated by a | so we simply have to call the split function on |
movies_df['genres'] = movies_df.genres.str.split('|')
movies_df.head()

movieId                        title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   

                                              genres  year  
0  [Adventure, Animation, Children, Comedy, Fantasy]  1995  
1                     [Adventure, Children, Fantasy]  1995  
2                                  [Comedy, Romance]  1995  
3                           [Comedy, Drama, Romance]  1995  
4                                           [Comedy]  1995

Since keeping genres in a list format isn't optimal for the content-based recommendation system technique, we will use the One Hot Encoding technique to convert the list of genres to a vector where each column corresponds to one possible value of the feature. This encoding is needed for feeding categorical data. In this case, we store every different genre in columns that contain either 1 or 0. 1 shows that a movie has that genre and 0 shows that it doesn't. Let's also store this dataframe in another variable since genres won't be important for our first recommendation system.


In [9]:
#Copying the movie dataframe into a new one since we won't need to use the genre information in our first case.
moviesWithGenres_df = movies_df.copy()

#For every row in the dataframe, iterate through the list of genres and place a 1 into the corresponding column
for index, row in movies_df.iterrows():
    for genre in row['genres']:
        moviesWithGenres_df.at[index, genre] = 1
#Filling in the NaN values with 0 to show that a movie doesn't have that column's genre
moviesWithGenres_df = moviesWithGenres_df.fillna(0)
moviesWithGenres_df.head()

movieId                        title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   

                                              genres  year  Adventure  \
0  [Adventure, Animation, Children, Comedy, Fantasy]  1995        1.0   
1                     [Adventure, Children, Fantasy]  1995        1.0   
2                                  [Comedy, Romance]  1995        0.0   
3                           [Comedy, Drama, Romance]  1995        0.0   
4                                           [Comedy]  1995        0.0   

   Animation  Children  Comedy  Fantasy  Romance  ...  Horror  Mystery  \
0        1.0       1.0     1.0      1.0      0.0  ...     0.0      0.0   
1        0.0       1.0     0.0      1.0      0.0  ...     0.0      0.0   
2        0.0       0.0     1.0      0.0      1.0  ...     0.0      0.0   
3        0.0       0.0     1.0      0.0      1.0  ...     0.0      0.0   
4        0.0       0.0     1.0      0.0      0.0  ...     0.0      0.0   

   Sci-Fi  IMAX  Documentary  War  Musical  Western  Film-Noir  \
0     0.0   0.0          0.0  0.0      0.0      0.0        0.0   
1     0.0   0.0          0.0  0.0      0.0      0.0        0.0   
2     0.0   0.0          0.0  0.0      0.0      0.0        0.0   
3     0.0   0.0          0.0  0.0      0.0      0.0        0.0   
4     0.0   0.0          0.0  0.0      0.0      0.0        0.0   

   (no genres listed)  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0  

[5 rows x 24 columns]

Next, let's look at the ratings dataframe.


In [10]:
ratings_df.head()

userId  movieId  rating   timestamp
0       1      169     2.5  1204927694
1       1     2471     3.0  1204927438
2       1    48516     5.0  1204927435
3       2     2571     3.5  1436165433
4       2   109487     4.0  1436165496

Every row in the ratings dataframe has a user id associated with at least one movie, a rating and a timestamp showing when they reviewed it. We won't be needing the timestamp column, so let's drop it to save memory.


In [11]:
#Drop removes a specified row or column from a dataframe
ratings_df = ratings_df.drop('timestamp', 1)
ratings_df.head()

C:\Users\admin\AppData\Local\Temp\ipykernel_18044\3391429438.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  ratings_df = ratings_df.drop('timestamp', 1)


userId  movieId  rating
0       1      169     2.5
1       1     2471     3.0
2       1    48516     5.0
3       2     2571     3.5
4       2   109487     4.0

<a id="ref3"></a>

# Content-Based recommendation system


Now, let's take a look at how to implement **Content-Based** or **Item-Item recommendation systems**. This technique attempts to figure out what a user's favourite aspects of an item is, and then recommends items that present those aspects. In our case, we're going to try to figure out the input's favorite genres from the movies and ratings given.

Let's begin by creating an input user to recommend movies to:

Notice: To add more movies, simply increase the amount of elements in the **userInput**. Feel free to add more in! Just be sure to write it in with capital letters and if a movie starts with a "The", like "The Matrix" then write it in like this: 'Matrix, The' .


In [15]:
userInput = [
            {'title':'Breakfast Club, The', 'rating':5},
            {'title':'Toy Story', 'rating':3.5},
            {'title':'Jumanji', 'rating':2},
            {'title':"Pulp Fiction", 'rating':5},
            {'title':'Akira', 'rating':4.5},
            {'title': 'Karate Kid, The', 'rating':5},
         ] 
inputMovies = pd.DataFrame(userInput)
inputMovies

title  rating
0  Breakfast Club, The     5.0
1            Toy Story     3.5
2              Jumanji     2.0
3         Pulp Fiction     5.0
4                Akira     4.5
5      Karate Kid, The     5.0

#### Add movieId to input user

With the input complete, let's extract the input movie's ID's from the movies dataframe and add them into it.

We can achieve this by first filtering out the rows that contain the input movie's title and then merging this subset with the input dataframe. We also drop unnecessary columns for the input to save memory space.


In [16]:
#Filtering out the movies by title
inputId = movies_df[movies_df['title'].isin(inputMovies['title'].tolist())]
#Then merging it so we can get the movieId. It's implicitly merging it by title.
inputMovies = pd.merge(inputId, inputMovies)
#Dropping information we won't use from the input dataframe
inputMovies = inputMovies.drop('genres', 1).drop('year', 1)
#Final input dataframe
#If a movie you added in above isn't here, then it might not be in the original 
#dataframe or it might spelled differently, please check capitalisation.
inputMovies

C:\Users\admin\AppData\Local\Temp\ipykernel_18044\2071048360.py:6: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  inputMovies = inputMovies.drop('genres', 1).drop('year', 1)
C:\Users\admin\AppData\Local\Temp\ipykernel_18044\2071048360.py:6: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  inputMovies = inputMovies.drop('genres', 1).drop('year', 1)


movieId                title  rating
0        1            Toy Story     3.5
1        2              Jumanji     2.0
2      296         Pulp Fiction     5.0
3     1274                Akira     4.5
4     1968  Breakfast Club, The     5.0
5     2420      Karate Kid, The     5.0
6    79224      Karate Kid, The     5.0

We're going to start by learning the input's preferences, so let's get the subset of movies that the input has watched from the Dataframe containing genres defined with binary values.


In [17]:
#Filtering out the movies from the input
userMovies = moviesWithGenres_df[moviesWithGenres_df['movieId'].isin(inputMovies['movieId'].tolist())]
userMovies

movieId                title  \
0            1            Toy Story   
1            2              Jumanji   
293        296         Pulp Fiction   
1246      1274                Akira   
1885      1968  Breakfast Club, The   
2336      2420      Karate Kid, The   
15579    79224      Karate Kid, The   

                                                  genres  year  Adventure  \
0      [Adventure, Animation, Children, Comedy, Fantasy]  1995        1.0   
1                         [Adventure, Children, Fantasy]  1995        1.0   
293                     [Comedy, Crime, Drama, Thriller]  1994        0.0   
1246              [Action, Adventure, Animation, Sci-Fi]  1988        1.0   
1885                                     [Comedy, Drama]  1985        0.0   
2336                                             [Drama]  1984        0.0   
15579                          [Action, Children, Drama]  2010        0.0   

       Animation  Children  Comedy  Fantasy  Romance  ...  Horror  Mystery  \
0            1.0       1.0     1.0      1.0      0.0  ...     0.0      0.0   
1            0.0       1.0     0.0      1.0      0.0  ...     0.0      0.0   
293          0.0       0.0     1.0      0.0      0.0  ...     0.0      0.0   
1246         1.0       0.0     0.0      0.0      0.0  ...     0.0      0.0   
1885         0.0       0.0     1.0      0.0      0.0  ...     0.0      0.0   
2336         0.0       0.0     0.0      0.0      0.0  ...     0.0      0.0   
15579        0.0       1.0     0.0      0.0      0.0  ...     0.0      0.0   

       Sci-Fi  IMAX  Documentary  War  Musical  Western  Film-Noir  \
0         0.0   0.0          0.0  0.0      0.0      0.0        0.0   
1         0.0   0.0          0.0  0.0      0.0      0.0        0.0   
293       0.0   0.0          0.0  0.0      0.0      0.0        0.0   
1246      1.0   0.0          0.0  0.0      0.0      0.0        0.0   
1885      0.0   0.0          0.0  0.0      0.0      0.0        0.0   
2336      0.0   0.0          0.0  0.0      0.0      0.0        0.0   
15579     0.0   0.0          0.0  0.0      0.0      0.0        0.0   

       (no genres listed)  
0                     0.0  
1                     0.0  
293                   0.0  
1246                  0.0  
1885                  0.0  
2336                  0.0  
15579                 0.0  

[7 rows x 24 columns]

We'll only need the actual genre table, so let's clean this up a bit by resetting the index and dropping the movieId, title, genres and year columns.


In [18]:
#Resetting the index to avoid future issues
userMovies = userMovies.reset_index(drop=True)
#Dropping unnecessary issues due to save memory and to avoid issues
userGenreTable = userMovies.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
userGenreTable

C:\Users\admin\AppData\Local\Temp\ipykernel_18044\2641803640.py:4: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  userGenreTable = userMovies.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
C:\Users\admin\AppData\Local\Temp\ipykernel_18044\2641803640.py:4: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  userGenreTable = userMovies.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
C:\Users\admin\AppData\Local\Temp\ipykernel_18044\2641803640.py:4: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  userGenreTable = userMovies.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
C:\Users\admin\AppData\Local\Temp\ipykernel_18044\2641803640.py:4: FutureWarning: In a future version o

Adventure  Animation  Children  Comedy  Fantasy  Romance  Drama  Action  \
0        1.0        1.0       1.0     1.0      1.0      0.0    0.0     0.0   
1        1.0        0.0       1.0     0.0      1.0      0.0    0.0     0.0   
2        0.0        0.0       0.0     1.0      0.0      0.0    1.0     0.0   
3        1.0        1.0       0.0     0.0      0.0      0.0    0.0     1.0   
4        0.0        0.0       0.0     1.0      0.0      0.0    1.0     0.0   
5        0.0        0.0       0.0     0.0      0.0      0.0    1.0     0.0   
6        0.0        0.0       1.0     0.0      0.0      0.0    1.0     1.0   

   Crime  Thriller  Horror  Mystery  Sci-Fi  IMAX  Documentary  War  Musical  \
0    0.0       0.0     0.0      0.0     0.0   0.0          0.0  0.0      0.0   
1    0.0       0.0     0.0      0.0     0.0   0.0          0.0  0.0      0.0   
2    1.0       1.0     0.0      0.0     0.0   0.0          0.0  0.0      0.0   
3    0.0       0.0     0.0      0.0     1.0   0.0          0.0  0.0      0.0   
4    0.0       0.0     0.0      0.0     0.0   0.0          0.0  0.0      0.0   
5    0.0       0.0     0.0      0.0     0.0   0.0          0.0  0.0      0.0   
6    0.0       0.0     0.0      0.0     0.0   0.0          0.0  0.0      0.0   

   Western  Film-Noir  (no genres listed)  
0      0.0        0.0                 0.0  
1      0.0        0.0                 0.0  
2      0.0        0.0                 0.0  
3      0.0        0.0                 0.0  
4      0.0        0.0                 0.0  
5      0.0        0.0                 0.0  
6      0.0        0.0                 0.0

Now we're ready to start learning the input's preferences!

To do this, we're going to turn each genre into weights. We can do this by using the input's reviews and multiplying them into the input's genre table and then summing up the resulting table by column. This operation is actually a dot product between a matrix and a vector, so we can simply accomplish by calling the Pandas "dot" function.


In [19]:
inputMovies['rating']

0    3.5
1    2.0
2    5.0
3    4.5
4    5.0
5    5.0
6    5.0
Name: rating, dtype: float64

In [20]:
#Dot produt to get weights
userProfile = userGenreTable.transpose().dot(inputMovies['rating'])
#The user profile
userProfile

Adventure             10.0
Animation              8.0
Children              10.5
Comedy                13.5
Fantasy                5.5
Romance                0.0
Drama                 20.0
Action                 9.5
Crime                  5.0
Thriller               5.0
Horror                 0.0
Mystery                0.0
Sci-Fi                 4.5
IMAX                   0.0
Documentary            0.0
War                    0.0
Musical                0.0
Western                0.0
Film-Noir              0.0
(no genres listed)     0.0
dtype: float64

Now, we have the weights for every of the user's preferences. This is known as the User Profile. Using this, we can recommend movies that satisfy the user's preferences.


Let's start by extracting the genre table from the original dataframe:


In [31]:
#Now let's get the genres of every movie in our original dataframe
genreTable = moviesWithGenres_df.set_index(moviesWithGenres_df['movieId'])
#And drop the unnecessary information
genreTable = genreTable.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
genreTable

C:\Users\admin\AppData\Local\Temp\ipykernel_18044\4125877002.py:4: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  genreTable = genreTable.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
C:\Users\admin\AppData\Local\Temp\ipykernel_18044\4125877002.py:4: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  genreTable = genreTable.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
C:\Users\admin\AppData\Local\Temp\ipykernel_18044\4125877002.py:4: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  genreTable = genreTable.drop('movieId', 1).drop('title', 1).drop('genres', 1).drop('year', 1)
C:\Users\admin\AppData\Local\Temp\ipykernel_18044\4125877002.py:4: FutureWarning: In a future version of pandas all

Adventure  Animation  Children  Comedy  Fantasy  Romance  Drama  \
movieId                                                                    
1              1.0        1.0       1.0     1.0      1.0      0.0    0.0   
2              1.0        0.0       1.0     0.0      1.0      0.0    0.0   
3              0.0        0.0       0.0     1.0      0.0      1.0    0.0   
4              0.0        0.0       0.0     1.0      0.0      1.0    1.0   
5              0.0        0.0       0.0     1.0      0.0      0.0    0.0   
...            ...        ...       ...     ...      ...      ...    ...   
151697         0.0        0.0       0.0     0.0      0.0      0.0    0.0   
151701         0.0        0.0       0.0     0.0      0.0      0.0    0.0   
151703         0.0        0.0       0.0     0.0      0.0      0.0    1.0   
151709         0.0        0.0       0.0     0.0      0.0      0.0    1.0   
151711         0.0        0.0       0.0     0.0      0.0      0.0    0.0   

         Action  Crime  Thriller  Horror  Mystery  Sci-Fi  IMAX  Documentary  \
movieId                                                                        
1           0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   
2           0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   
3           0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   
4           0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   
5           0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   
...         ...    ...       ...     ...      ...     ...   ...          ...   
151697      0.0    0.0       1.0     0.0      0.0     0.0   0.0          0.0   
151701      0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   
151703      0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   
151709      0.0    0.0       0.0     0.0      0.0     1.0   0.0          0.0   
151711      0.0    0.0       0.0     0.0      0.0     0.0   0.0          0.0   

         War  Musical  Western  Film-Noir  (no genres listed)  
movieId                                                        
1        0.0      0.0      0.0        0.0                 0.0  
2        0.0      0.0      0.0        0.0                 0.0  
3        0.0      0.0      0.0        0.0                 0.0  
4        0.0      0.0      0.0        0.0                 0.0  
5        0.0      0.0      0.0        0.0                 0.0  
...      ...      ...      ...        ...                 ...  
151697   0.0      0.0      0.0        0.0                 0.0  
151701   0.0      0.0      0.0        0.0                 1.0  
151703   0.0      0.0      0.0        0.0                 0.0  
151709   0.0      0.0      0.0        0.0                 0.0  
151711   0.0      0.0      0.0        0.0                 1.0  

[34208 rows x 20 columns]

In [27]:
print(genreTable.shape)
print(userProfile.shape)

(34208, 20)
(20,)


With the input's profile and the complete list of movies and their genres in hand, we're going to take the weighted average of every movie based on the input profile and recommend the top twenty movies that most satisfy it.


In [30]:
#Multiply the genres by the weights and then take the weighted average
recommendationTable_df = ((genreTable*userProfile).sum(axis=1))/(userProfile.sum())
recommendationTable_df

movieId
1         0.519126
2         0.284153
3         0.147541
4         0.366120
5         0.147541
            ...   
151697    0.054645
151701    0.000000
151703    0.218579
151709    0.267760
151711    0.000000
Length: 34208, dtype: float64

In [32]:
#Sort our recommendations in descending order
recommendationTable_df = recommendationTable_df.sort_values(ascending=False)
#Just a peek at the values
recommendationTable_df

movieId
26093     0.737705
27344     0.726776
148775    0.699454
149488    0.699454
5018      0.693989
            ...   
82143     0.000000
133225    0.000000
133221    0.000000
82157     0.000000
151711    0.000000
Length: 34208, dtype: float64

Now here's the recommendation table!


In [33]:
#The final recommendation table
movies_df.loc[movies_df['movieId'].isin(recommendationTable_df.head(20).keys())]

movieId                                              title  \
1824      1907                                              Mulan   
4625      4719                                      Osmosis Jones   
4923      5018                                           Motorama   
8605     26093         Wonderful World of the Brothers Grimm, The   
9296     27344  Revolutionary Girl Utena: Adolescence of Utena...   
13250    64645                                  The Wrecking Crew   
15001    75408  Lupin III: Sweet Lost Night (Rupan Sansei: Swe...   
15073    76153  Lupin III: First Contact (Rupan Sansei: Faasut...   
16055    81132                                             Rubber   
18312    91335                                      Gruffalo, The   
22778   108540           Ernest & Célestine (Ernest et Célestine)   
25218   117646                     Dragonheart 2: A New Beginning   
26442   122787                                       The 39 Steps   
26563   123304                                            Scrooge   
26770   124519                   Snow White and the Three Stooges   
27097   125972                Halloweentown II: Kalabar's Revenge   
32313   144350                                 Under the Mountain   
32854   146305                             Princes and Princesses   
33509   148775                Wizards of Waverly Place: The Movie   
33692   149488                                     Christmas Town   

                                                  genres  year  
1824   [Adventure, Animation, Children, Comedy, Drama...  1998  
4625   [Action, Animation, Comedy, Crime, Drama, Roma...  2001  
4923   [Adventure, Comedy, Crime, Drama, Fantasy, Mys...  1991  
8605   [Adventure, Animation, Children, Comedy, Drama...  1962  
9296   [Action, Adventure, Animation, Comedy, Drama, ...  1999  
13250  [Action, Adventure, Comedy, Crime, Drama, Thri...  1968  
15001  [Action, Animation, Comedy, Crime, Drama, Myst...  2008  
15073  [Action, Animation, Comedy, Crime, Drama, Myst...  2002  
16055  [Action, Adventure, Comedy, Crime, Drama, Film...  2010  
18312    [Adventure, Animation, Children, Comedy, Drama]  2009  
22778  [Adventure, Animation, Children, Comedy, Drama...  2012  
25218  [Action, Adventure, Comedy, Drama, Fantasy, Th...  2000  
26442  [Action, Adventure, Comedy, Crime, Drama, Thri...  1959  
26563         [Action, Children, Comedy, Drama, Fantasy]  1935  
26770      [Adventure, Children, Comedy, Drama, Fantasy]  1961  
27097      [Adventure, Children, Comedy, Drama, Fantasy]  2001  
32313  [Action, Adventure, Children, Drama, Fantasy, ...  2009  
32854  [Animation, Children, Comedy, Drama, Fantasy, ...  2000  
33509  [Adventure, Children, Comedy, Drama, Fantasy, ...  2009  
33692  [Action, Children, Comedy, Drama, Fantasy, Thr...  2008

### Advantages and Disadvantages of Content-Based Filtering

##### Advantages

*   Learns user's preferences
*   Highly personalized for the user

##### Disadvantages

*   Doesn't take into account what others think of the item, so low quality item recommendations might happen
*   Extracting data is not always intuitive
*   Determining what characteristics of the item the user dislikes or likes is not always obvious


<h2>Want to learn more?</h2>

IBM SPSS Modeler is a comprehensive analytics platform that has many machine learning algorithms. It has been designed to bring predictive intelligence to decisions made by individuals, by groups, by systems – by your enterprise as a whole. A free trial is available through this course, available here: <a href="https://www.ibm.com/analytics/spss-statistics-software?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01">SPSS Modeler</a>

Also, you can use Watson Studio to run these notebooks faster with bigger datasets. Watson Studio is IBM's leading cloud solution for data scientists, built by data scientists. With Jupyter notebooks, RStudio, Apache Spark and popular libraries pre-packaged in the cloud, Watson Studio enables data scientists to collaborate on their projects without having to install anything. Join the fast-growing community of Watson Studio users today with a free account at <a href="https://www.ibm.com/cloud/watson-studio?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01">Watson Studio</a>


### Thank you for completing this lab!

## Author

Saeed Aghabozorgi

### Other Contributors

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01" target="_blank">Joseph Santarcangelo</a>

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                 |
| ----------------- | ------- | ---------- | ---------------------------------- |
| 2020-11-03        | 2.1     | Lakshmi    | Updated URL of csv                 |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab |
|                   |         |            |                                    |
|                   |         |            |                                    |

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
